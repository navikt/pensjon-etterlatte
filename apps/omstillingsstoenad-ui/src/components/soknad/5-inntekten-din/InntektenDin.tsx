import SoknadSteg from '../../../typer/SoknadSteg'
import { SkjemaGruppe } from '../../felles/SkjemaGruppe'
import { FormProvider, useForm } from 'react-hook-form'
import { ActionTypes } from '../../../context/soknad/soknad'
import { useSoknadContext } from '../../../context/soknad/SoknadContext'
import Feilmeldinger from '../../felles/Feilmeldinger'
import Navigasjon from '../../felles/Navigasjon'
import { useTranslation } from 'react-i18next'
import { GuidePanel, Heading } from '@navikt/ds-react'
import { deepCopy } from '../../../utils/deepCopy'
import { SkjemaElement } from '../../felles/SkjemaElement'
import { IInntekt } from '../../../typer/inntekt'

const InntektenDin: SoknadSteg = ({ neste, forrige }) => {
    const { t } = useTranslation()

    const { state, dispatch } = useSoknadContext()

    const methods = useForm<IInntekt>({
        defaultValues: state.dinSituasjon || {},
        shouldUnregister: true,
    })

    const {
        handleSubmit,
        formState: { errors },
        getValues,
    } = methods

    const lagreNeste = (data: IInntekt) => {
        dispatch({ type: ActionTypes.OPPDATER_INNTEKTEN_DIN, payload: { ...deepCopy(data), erValidert: true } })
        neste!!()
    }

    const lagreTilbake = (data: IInntekt) => {
        dispatch({ type: ActionTypes.OPPDATER_INNTEKTEN_DIN, payload: { ...deepCopy(data), erValidert: true } })
        forrige!!()
    }

    const lagreTilbakeUtenValidering = () => {
        const verdier = getValues()
        dispatch({ type: ActionTypes.OPPDATER_INNTEKTEN_DIN, payload: { ...deepCopy(verdier), erValidert: false } })
        forrige!!()
    }

    const erValidert = state.inntektenDin.erValidert

    return (
        <FormProvider {...methods}>
            <form>
                <SkjemaElement>
                    <Heading size={'medium'} className={'center'}>
                        {t('inntektenDin.tittel')}
                    </Heading>
                </SkjemaElement>

                <SkjemaGruppe>
                    <GuidePanel>
                        <Heading size={'small'}>{t('dinSituasjon.undertittel')}</Heading>
                        {t('dinSituasjon.ingress')}
                    </GuidePanel>
                </SkjemaGruppe>

                <Feilmeldinger errors={errors} />

                <Navigasjon
                    forrige={{ onClick: erValidert === true ? handleSubmit(lagreTilbake) : lagreTilbakeUtenValidering }}
                    neste={{ onClick: handleSubmit(lagreNeste) }}
                />
            </form>
        </FormProvider>
    )
}

export default InntektenDin
