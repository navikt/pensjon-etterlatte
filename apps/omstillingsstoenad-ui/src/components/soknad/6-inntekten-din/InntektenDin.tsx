import { Box, GuidePanel, Heading, VStack } from '@navikt/ds-react'
import { differenceInYears } from 'date-fns'
import { FormProvider, useForm } from 'react-hook-form'
import { FieldErrors } from 'react-hook-form/dist/types/errors'
import { useTranslation } from 'react-i18next'
import { useBrukerContext } from '~context/bruker/BrukerContext'
import { FeatureToggleNavn, FeatureToggleStatus, useFeatureToggle } from '~context/featureToggle/FeatureToggleContext'
import { LogEvents, useAnalytics } from '~hooks/useAnalytics'
import { IInntekt } from '~typer/inntekt'
import { erMellomOktoberogDesember } from '~utils/dato'
import { useSoknadContext } from '../../../context/soknad/SoknadContext'
import { ActionTypes } from '../../../context/soknad/soknad'
import SoknadSteg from '../../../typer/SoknadSteg'
import { deepCopy } from '../../../utils/deepCopy'
import Feilmeldinger from '../../felles/Feilmeldinger'
import Navigasjon from '../../felles/Navigasjon'
import { ForventetInntektIAar } from './fragmenter/ForventetInntektIAar'
import { ForventetInntektTilNesteAar } from './fragmenter/ForventetInntektTilNesteAar'
import Inntekt from './fragmenter/Inntekt'
import { InntektFremTilDoedsfallet } from './fragmenter/InntektFremTilDoedsfallet'
import { SkalGaaAvMedAlderspensjon } from './fragmenter/SkalGaaAvMedAlderspensjon'
import YtelserAndre from './fragmenter/YtelserAndre'
import YtelserNAV from './fragmenter/YtelserNAV'

const InntektenDin = ({ neste, forrige }: SoknadSteg) => {
    const { t } = useTranslation()
    const { logEvent } = useAnalytics()

    const omsSoeknadNyttInntektStegFeatureToggle = useFeatureToggle(FeatureToggleNavn.OMS_SOEKNAD_NYTT_INNTEKT_STEG)

    const { state, dispatch } = useSoknadContext()

    const { state: bruker } = useBrukerContext()

    const methods = useForm<IInntekt>({
        defaultValues: state.inntektenDin || {},
        shouldUnregister: true,
    })

    const {
        handleSubmit,
        formState: { errors },
        getValues,
        watch,
    } = methods

    const lagreNeste = (data: IInntekt) => {
        dispatch({ type: ActionTypes.OPPDATER_INNTEKTEN_DIN, payload: { ...deepCopy(data), erValidert: true } })
        neste!()
    }

    const lagreTilbake = (data: IInntekt) => {
        dispatch({ type: ActionTypes.OPPDATER_INNTEKTEN_DIN, payload: { ...deepCopy(data), erValidert: true } })
        forrige!()
    }

    const lagreTilbakeUtenValidering = () => {
        const verdier = getValues()
        dispatch({ type: ActionTypes.OPPDATER_INNTEKTEN_DIN, payload: { ...deepCopy(verdier), erValidert: false } })
        forrige!()
    }

    const logErrors = (data: FieldErrors<IInntekt>) => {
        Object.keys(data).map((error) =>
            logEvent(LogEvents.VALIDATION_ERROR, { skjemanavn: 'InntektenDin', id: error })
        )
    }

    const erValidert = state.inntektenDin.erValidert

    const skalViseFelterForSkalGaaAvMedAlderspensjon = (): boolean => {
        if (bruker.foedselsdato) {
            const alder = differenceInYears(new Date(), bruker.foedselsdato)
            return alder >= 62
        }

        return false
    }

    return (
        <FormProvider {...methods}>
            <form onSubmit={(e) => e.preventDefault()}>
                <Box marginBlock="4">
                    <Heading size={'medium'} className={'center'}>
                        {t('inntektenDin.tittel')}
                    </Heading>
                </Box>

                <Box marginBlock="0 12">
                    <GuidePanel>
                        {t(
                            omsSoeknadNyttInntektStegFeatureToggle.status === FeatureToggleStatus.PAA
                                ? 'inntektenDin.ingress.ny'
                                : 'inntektenDin.ingress'
                        )}
                    </GuidePanel>
                </Box>

                {omsSoeknadNyttInntektStegFeatureToggle.status === FeatureToggleStatus.PAA ? (
                    <VStack gap="12" paddingBlock="0 12">
                        {skalViseFelterForSkalGaaAvMedAlderspensjon() ? (
                            <>
                                <SkalGaaAvMedAlderspensjon />
                                {!!watch('skalGaaAvMedAlderspensjon.valg') && (
                                    <>
                                        <VStack gap="6">
                                            <Heading size="medium">{t('inntektenDin.inntekteneDine.tittel')}</Heading>

                                            <InntektFremTilDoedsfallet />
                                        </VStack>

                                        <ForventetInntektIAar />

                                        {erMellomOktoberogDesember() && <ForventetInntektTilNesteAar />}
                                    </>
                                )}
                            </>
                        ) : (
                            <>
                                <VStack gap="6">
                                    <Heading size="medium">{t('inntektenDin.inntekteneDine.tittel')}</Heading>

                                    <InntektFremTilDoedsfallet />
                                </VStack>

                                <ForventetInntektIAar />

                                {erMellomOktoberogDesember() && <ForventetInntektTilNesteAar />}
                            </>
                        )}
                    </VStack>
                ) : (
                    <Inntekt />
                )}

                <YtelserNAV />

                <YtelserAndre />

                <Feilmeldinger errors={errors} />

                <Navigasjon
                    forrige={{
                        onClick:
                            erValidert === true ? handleSubmit(lagreTilbake, logErrors) : lagreTilbakeUtenValidering,
                    }}
                    neste={{ onClick: handleSubmit(lagreNeste, logErrors) }}
                />
            </form>
        </FormProvider>
    )
}

export default InntektenDin
