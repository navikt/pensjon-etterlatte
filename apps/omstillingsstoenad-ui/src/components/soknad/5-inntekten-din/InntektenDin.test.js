import { act, render, fireEvent } from '@testing-library/react'
import InntektenDin from './InntektenDin'

jest.mock('react-i18next', () => ({
    // this mock makes sure any components using the translate hook can use it without a warning being shown
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str) => str,
            i18n: {
                changeLanguage: () => new Promise(() => {}),
            },
        }
    },
}))

const defaultValues = {}

jest.mock('../../../context/soknad/SoknadContext', () => ({
    useSoknadContext: () => ({
        state: { inntektenDin: defaultValues },
        dispatch: jest.fn(),
    }),
}))

describe('Inntekten din', () => {
    it('Snapshot', () => {
        const { container } = render(<InntektenDin />)
        expect(container).toMatchSnapshot()
    })

    it('Skal rendre selvstendig', () => {
        const { container, getByLabelText } = render(<InntektenDin />)
        /*act(() => {
            fireEvent.click(getByLabelText('jobbStatus.selvstendig'))
        })*/
        //console.log(container.querySelectorAll("input[name=selvstendig[0].forventerEndretInntekt.svar]"))
        /*
        act(() => {
           fireEvent.change(getByText("dinSituasjon.selvstendig.forventerEndretInntekt.svar"), { target: { value: "Ja" } });
        });
        */
        //expect(getByLabelText("dinSituasjon.selvstendig.forventerEndretInntekt.beskrivelse")).toBeDefined();
        expect(container).toMatchSnapshot()
    })
})
