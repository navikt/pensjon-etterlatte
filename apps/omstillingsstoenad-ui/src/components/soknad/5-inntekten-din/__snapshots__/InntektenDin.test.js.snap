// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inntekten din Skal rendre selvstendig 1`] = `
<div>
  <form>
    <div
      class="sc-irLvoH eNFBNd"
    >
      <h1
        class="center navds-heading navds-heading--medium"
      >
        inntektenDin.tittel
      </h1>
    </div>
    <fieldset
      class="sc-beyTiQ jltUnl"
    >
      <h1
        class="navds-heading navds-heading--small"
      >
        dinSituasjon.undertittel
      </h1>
      <p
        class="navds-body-long navds-body-long--medium"
      >
        dinSituasjon.ingress
      </p>
    </fieldset>
    <fieldset
      class="sc-beyTiQ jltUnl"
    >
      <div
        class="sc-guDMob czmIqC undefined"
      >
        <button
          class="navds-button navds-button--primary navds-button--medium"
          type="button"
        >
          <span
            class="navds-label"
          >
            knapp.tilbake
          </span>
        </button>
        <button
          class="navds-button navds-button--primary navds-button--medium"
          type="button"
        >
          <span
            class="navds-label"
          >
            knapp.neste
          </span>
        </button>
      </div>
    </fieldset>
    <div
      class="sc-guDMob sc-dmyDmy czmIqC kgmmuh"
    >
      <button
        class="navds-button navds-button--secondary navds-button--medium"
        id="avbryt-btn"
        type="button"
      >
        <span
          class="navds-label"
        >
          knapp.avbryt
        </span>
      </button>
    </div>
    <dialog
      class="navds-modal navds-modal--polyfilled navds-modal--autowidth"
      role="dialog"
    >
      <div
        class="navds-modal__header"
      >
        <button
          class="navds-modal__button navds-button navds-button--tertiary-neutral navds-button--small navds-button--icon-only"
          type="button"
        >
          <span
            class="navds-button__icon"
          >
            <svg
              aria-labelledby="title-r5"
              fill="none"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="title-r5"
              >
                Lukk modalvindu
              </title>
              <path
                clip-rule="evenodd"
                d="M6.53 5.47a.75.75 0 0 0-1.06 1.06L10.94 12l-5.47 5.47a.75.75 0 1 0 1.06 1.06L12 13.06l5.47 5.47a.75.75 0 1 0 1.06-1.06L13.06 12l5.47-5.47a.75.75 0 0 0-1.06-1.06L12 10.94 6.53 5.47Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
        <h1
          class="navds-heading navds-heading--medium"
        >
          avbrytModal.spoersmaal
        </h1>
      </div>
      <div
        class="navds-modal__body"
      >
        <p
          class="mute avbryt-text navds-body-short navds-body-short--medium"
        >
          avbrytModal.informasjon
        </p>
      </div>
      <div
        class="navds-modal__footer"
      >
        <div
          class="sc-jsEeA-d gHwiqO"
        >
          <button
            class="navds-button navds-button--secondary navds-button--medium"
            id="avbryt-nei-btn"
            style="margin: 10px;"
            type="button"
          >
            <span
              class="navds-label"
            >
              avbrytModal.svarNei
            </span>
          </button>
          <button
            class="navds-button navds-button--primary navds-button--medium"
            id="avbryt-ja-btn"
            style="margin: 10px;"
            type="button"
          >
            <span
              class="navds-label"
            >
              avbrytModal.svarJa
            </span>
          </button>
        </div>
        <div
          class="sc-jsEeA-d gHwiqO"
        >
          <button
            class="navds-button navds-button--tertiary navds-button--medium"
            id="slett-soeknad"
            style="color: rgb(198, 93, 78);"
            type="button"
          >
            <span
              class="navds-label"
            >
              avbrytModal.svarSlett
            </span>
          </button>
        </div>
      </div>
    </dialog>
  </form>
</div>
`;

exports[`Inntekten din Snapshot 1`] = `
<div>
  <form>
    <div
      class="sc-irLvoH eNFBNd"
    >
      <h1
        class="center navds-heading navds-heading--medium"
      >
        inntektenDin.tittel
      </h1>
    </div>
    <fieldset
      class="sc-beyTiQ jltUnl"
    >
      <h1
        class="navds-heading navds-heading--small"
      >
        dinSituasjon.undertittel
      </h1>
      <p
        class="navds-body-long navds-body-long--medium"
      >
        dinSituasjon.ingress
      </p>
    </fieldset>
    <fieldset
      class="sc-beyTiQ jltUnl"
    >
      <div
        class="sc-guDMob czmIqC undefined"
      >
        <button
          class="navds-button navds-button--primary navds-button--medium"
          type="button"
        >
          <span
            class="navds-label"
          >
            knapp.tilbake
          </span>
        </button>
        <button
          class="navds-button navds-button--primary navds-button--medium"
          type="button"
        >
          <span
            class="navds-label"
          >
            knapp.neste
          </span>
        </button>
      </div>
    </fieldset>
    <div
      class="sc-guDMob sc-dmyDmy czmIqC kgmmuh"
    >
      <button
        class="navds-button navds-button--secondary navds-button--medium"
        id="avbryt-btn"
        type="button"
      >
        <span
          class="navds-label"
        >
          knapp.avbryt
        </span>
      </button>
    </div>
    <dialog
      class="navds-modal navds-modal--polyfilled navds-modal--autowidth"
      role="dialog"
    >
      <div
        class="navds-modal__header"
      >
        <button
          class="navds-modal__button navds-button navds-button--tertiary-neutral navds-button--small navds-button--icon-only"
          type="button"
        >
          <span
            class="navds-button__icon"
          >
            <svg
              aria-labelledby="title-r2"
              fill="none"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="title-r2"
              >
                Lukk modalvindu
              </title>
              <path
                clip-rule="evenodd"
                d="M6.53 5.47a.75.75 0 0 0-1.06 1.06L10.94 12l-5.47 5.47a.75.75 0 1 0 1.06 1.06L12 13.06l5.47 5.47a.75.75 0 1 0 1.06-1.06L13.06 12l5.47-5.47a.75.75 0 0 0-1.06-1.06L12 10.94 6.53 5.47Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
        <h1
          class="navds-heading navds-heading--medium"
        >
          avbrytModal.spoersmaal
        </h1>
      </div>
      <div
        class="navds-modal__body"
      >
        <p
          class="mute avbryt-text navds-body-short navds-body-short--medium"
        >
          avbrytModal.informasjon
        </p>
      </div>
      <div
        class="navds-modal__footer"
      >
        <div
          class="sc-jsEeA-d gHwiqO"
        >
          <button
            class="navds-button navds-button--secondary navds-button--medium"
            id="avbryt-nei-btn"
            style="margin: 10px;"
            type="button"
          >
            <span
              class="navds-label"
            >
              avbrytModal.svarNei
            </span>
          </button>
          <button
            class="navds-button navds-button--primary navds-button--medium"
            id="avbryt-ja-btn"
            style="margin: 10px;"
            type="button"
          >
            <span
              class="navds-label"
            >
              avbrytModal.svarJa
            </span>
          </button>
        </div>
        <div
          class="sc-jsEeA-d gHwiqO"
        >
          <button
            class="navds-button navds-button--tertiary navds-button--medium"
            id="slett-soeknad"
            style="color: rgb(198, 93, 78);"
            type="button"
          >
            <span
              class="navds-label"
            >
              avbrytModal.svarSlett
            </span>
          </button>
        </div>
      </div>
    </dialog>
  </form>
</div>
`;
